Replace global oversampling with overridable per-viewport oversampling.

bruvzg opened the pull request:
- Allows independent per-viewport oversampling (required for automatic DPI scaling). Previously was global value and only truly supported for the main window.
- Allows overriding oversampling factor manually.
- Oversampling can reuse existing font data if possible (e.g, if you have 16 and 32 size fonts already in use, and oversampling of 2x, oversampled 16 font will reuse 32 font data).
- Change of oversampling factor won't reload all fonts.
- Oversampling use 16.6 fractions units for font size (have more scaling steps, and won't reload font if scaling change is too small).
- ~~Adds performance monitors for `TextServer` fonts and oversampling.~~

bruvzg committed:
Replace global oversampling with overrideable per-viewport oversampling.

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/a8598cd8e261716fa3addb6f10bb57c03a061be9), I noticed an issue with embolden using a radius that's way too large in the editor. You can reproduce this by using `push_error()`/`push_warning()` in a `@tool` script's `_ready()` method, then using **Scene > Reload Saved Scene**:

![image](https://github.com/user-attachments/assets/8701cbe7-5fcc-4b72-a952-8c6f506edd50)

Also, this PR doesn't fix https://github.com/godotengine/godot/issues/86563:

Godot 3 | Godot 4 - `master` | Godot 4 - This PR
-|-|-
![Godot 3](https://github.com/user-attachments/assets/2a337594-3e5c-4575-9e8b-a81d0b6b8586) | ![Godot 4 Master](https://github.com/user-attachments/assets/5e7a7e80-f181-4b8d-bdb7-3930d28ede29) | ![Godot 4 PR](https://github.com/user-attachments/assets/5d3d298c-a797-4834-b391-e43c2c14fe7b)

At default window size, kerning has changed slightly on the text at the bottom-right corner, but the rest is identical:

Godot 3 | Godot 4 - `master` | Godot 4 - This PR
-|-|-
![Godot 3](https://github.com/user-attachments/assets/bb751679-257d-4f25-af06-6fa56285500a) | ![Godot 4 Master](https://github.com/user-attachments/assets/971b33a4-5280-40c7-ac97-c4a7f8ae5513) | ![Godot 4 PR](https://github.com/user-attachments/assets/6f211cc6-cac5-4dd5-baa5-af01d89e498c)

bruvzg commented:
Fixed embolden and issue with subpixel positioning activated at wrong size (was causing minor kerning shifts).

I'll check https://github.com/godotengine/godot/issues/86563.

clayjohn commented on servers/rendering/renderer_canvas_cull.h:
```diff
@@ -58,6 +58,7 @@ class RendererCanvasCull {
 		int ysort_index;
 		int ysort_parent_abs_z_index; // Absolute Z index of parent. Only populated and used when y-sorting.
 		uint32_t visibility_layer = 0xffffffff;
+		float oversampling_factor = 1.f;
```
Why are we storing this in the RenderingServer if the Renderer doesn't use it? Shouldn't this be stored on the same server/thread where it will actually be used?

bruvzg replied:
`TextServer` only have access to `CanvasItem` RID, so it can be either part of `RenderingServer` or passed as and extra text related draw call (which can be done, but involve a lot more changes).

clayjohn replied:
Sure, but retrieving the data from the RenderingServer is going to cause a stall every time it is called. Surely that is worse than having to pass the data around

clayjohn replied:
Actually, I don't even see `RS::canvas_item_get_oversampling_factor()` called anywhere. Is it even needed?

bruvzg replied:
It's called in `TextServer*::_font_draw_glyph` and `TextServer*::_font_draw_glyph_outline`.

bruvzg replied:
> stall every time it is called

It's a single RID owner access. And it's not showing in the Xcode time profiler at all.

Also, this makes oversampling transparent for the custom drawing by user, making it argument will require to do it manually.

clayjohn replied:
Look at the implementation of `FUNC1RC`, it syncs the CommandQueue every time it is called. 

For clarity, this only matters when the RenderingServer is on its own thread, but we need to be switching to that by default as soon as possible. If we add this now, we need to undo it later when we switch the Renderer to its own thread by default.

bruvzg replied:
I guess we'll need to keep local map of all canvas items in the TextServer, passing it as an argument will be a horrible mess.

Ideally there should be at least a way to get viewport RID from a CanvasItem RID.

> be switching to that by default as soon as possible

Is it even remotely functional, it was completely broken for a long time.

bruvzg replied:
Actually, I can just set current drawn item (since static `get_current_item_drawn()` exist I assume it is safe) oversampling from the `CanvasItem::_redraw_callback`, the same way it's currently updating it in the RS. I'll check it and remove RS changes in a moment.

bruvzg replied:
Removed it.

clayjohn replied:
> Is it even remotely functional, it was completely broken for a long time.

It is _remotely_ functional lol. The last thing we need to do is clean up all of these places that assume the RenderingServer will always be on the main thread.

KoBeWi commented on doc/classes/TextServer.xml:
```diff
@@ -391,11 +399,11 @@
 			<param index="0" name="font_rid" type="RID" />
 			<description>
-				Returns font oversampling factor, if set to [code]0.0[/code] global oversampling factor is used instead. Used by dynamic fonts only.
+				Deprecated. This method do nothing.
```
It does return something.

bruvzg replied:
Changed it to `This method always returns [code]1.0[/code].` to be more precise.

KoBeWi commented:
I added a Label in a SubViewport in a SubViewportContainer and setting `oversampling` doesn't have visible effect:

https://github.com/user-attachments/assets/111c0175-8ded-45c9-8f91-d7d39ac0a9dd

Also `oversampling_override` of 0 is somewhat buggy. When I set its value and then reset to 0, the override does not reset.

KoBeWi commented:
Shouldn't it have effect when I set the override though?

bruvzg commented:
> Shouldn't it have effect when I set the override though?

Yes, if override is set it always applies.

KoBeWi commented:
Then the `oversampling` property is misleading. It's difficult to tell when it has effect and when not. It should be clarified in the description, and the property can be disabled when irrelevant.

bruvzg commented:
Also, there's some max limit for absolute font size (4096 for single glyph dimensions), so if you set large font size and excessively large oversampling it might break (underlaying font size is multiple of these two).

KoBeWi commented on main/performance.cpp:
The new monitors look unrelated. Is it possible to split them into separate PR?

bruvzg replied:
Can be moved to https://github.com/godotengine/godot/pull/105083, to the rest of even more unrelated monitors.

KoBeWi commented on doc/classes/CanvasItem.xml:
```diff
@@ -56,6 +56,7 @@
 			<param index="2" name="char" type="String" />
 			<param index="3" name="font_size" type="int" default="16" />
 			<param index="4" name="modulate" type="Color" default="Color(1, 1, 1, 1)" />
+			<param index="5" name="oversampling" type="float" default="0.0" />
```
You could mention what the new argument does. I assume it's similar to the override in viewport.

bruvzg commented:
Added notes about new arguments and `Viewport.oversampling` usage and removed monitors (will move to https://github.com/godotengine/godot/pull/105083).

KoBeWi commented on modules/text_server_adv/text_server_adv.cpp:
```diff
@@ -3669,7 +3748,7 @@ PackedInt32Array TextServerAdvanced::_font_get_supported_glyphs(const RID &p_fon
 	FontForSizeAdvanced *at_size = nullptr;
 	if (fd->cache.is_empty()) {
-		ERR_FAIL_COND_V(!_ensure_cache_for_size(fd, fd->msdf ? Vector2i(fd->msdf_source_size, 0) : Vector2i(16, 0), at_size), PackedInt32Array());
+		ERR_FAIL_COND_V(!_ensure_cache_for_size(fd, fd->msdf ? Vector2i(fd->msdf_source_size * 64, 0) : Vector2i(16 * 64, 0), at_size), PackedInt32Array());
```
64 repeats quite a lot. Where does it come from? Can it be a constant?

bruvzg replied:
Convention to internal FreeType font size format (26.6 fractional units). It's using 32-bit integer as a fixed point real (26 bits real part, 6 bit fractional).

KoBeWi commented on scene/main/canvas_item.cpp:
```diff
@@ -136,15 +136,23 @@ void CanvasItem::_redraw_callback() {
+	RenderingServer::get_singleton()->canvas_item_clear(ci);
 	//todo updating = true - only allow drawing here
 	if (is_visible_in_tree()) {
+		Ref<TextServer> ts = TextServerManager::get_singleton()->get_primary_interface();
 		drawing = true;
+		if (ts.is_valid()) {
```
```suggestion
		drawing = true;
		Ref<TextServer> ts = TextServerManager::get_singleton()->get_primary_interface();
		if (ts.is_valid()) {
```

KoBeWi reviewed (approved):
It's a complex change, but overall looks fine.

AThousandShips commented on doc/classes/ResourceImporterDynamicFont.xml:
```diff
@@ -67,7 +67,7 @@
 		</member>
 		<member name="oversampling" type="float" setter="" getter="" default="0.0">
-			If set to a value greater than [code]0.0[/code], overrides the oversampling factor for the font. This can be used to render the font at a higher or lower resolution than intended without affecting its physical size. In most cases, this should be left at [code]0.0[/code].
+			Deprecated. This property do nothing.
```
```suggestion
			Deprecated. This property does nothing.
```

AThousandShips commented on doc/classes/SystemFont.xml:
```diff
@@ -58,8 +58,8 @@
-		<member name="oversampling" type="float" setter="set_oversampling" getter="get_oversampling" default="0.0">
-			Font oversampling factor, if set to [code]0.0[/code] global oversampling factor is used instead.
+		<member name="oversampling" type="float" setter="set_oversampling" getter="get_oversampling" deprecated="Use [code skip-lint]oversampling[/code] argument of the [code skip-lint]draw_*[/code] methods instead.">
+			Deprecated. This property do nothing.
```
```suggestion
			Deprecated. This property does nothing.
```

AThousandShips commented on doc/classes/TextServer.xml:
```diff
@@ -421,6 +429,13 @@
+			<return type="Dictionary[]" />
+			<param index="0" name="font_rid" type="RID" />
+			<description>
+				Returns font cache information, each entry contains following fields: [code]Vector2i size_px[/code] - font size in pixels, [code]float viewport_oversampling[/code] - viewport oversampling factor, [code]int glyphs[/code] - number of rendered glyphs, [code]int textures[/code] - number of used textures, [code]int textures_size[/code] - size of texture data in bytes.
```
```suggestion
				Returns font cache information, each entry contains the following fields: [code]Vector2i size_px[/code] - font size in pixels, [code]float viewport_oversampling[/code] - viewport oversampling factor, [code]int glyphs[/code] - number of rendered glyphs, [code]int textures[/code] - number of used textures, [code]int textures_size[/code] - size of texture data in bytes.
```

AThousandShips commented on doc/classes/TextServer.xml:
```diff
@@ -771,12 +786,11 @@
 			<description>
-				Sets oversampling factor, shared by all font in the TextServer.
-				[b]Note:[/b] This value can be automatically changed by display server.
+				Deprecated. This method do nothing.
```
```suggestion
				Deprecated. This method does nothing.
```

AThousandShips commented on doc/classes/TextServer.xml:
```diff
@@ -914,12 +928,12 @@
 			<param index="1" name="oversampling" type="float" />
 			<description>
-				Sets font oversampling factor, if set to [code]0.0[/code] global oversampling factor is used instead. Used by dynamic fonts only.
+				Deprecated. This method do nothing.
```
```suggestion
				Deprecated. This method does nothing.
```

AThousandShips commented on doc/classes/TextServerExtension.xml:
```diff
@@ -447,6 +456,14 @@
+			<param index="0" name="font_rid" type="RID" />
+			<description>
+				[b]Optional.[/b]
+				Returns font cache information, each entry contains following fields: [code]Vector2i size_px[/code] - font size in pixels, [code]float viewport_oversampling[/code] - viewport oversampling factor, [code]int glyphs[/code] - number of rendered glyphs, [code]int textures[/code] - number of used textures, [code]int textures_size[/code] - size of texture data in bytes.
```
```suggestion
				Returns font cache information, each entry contains the following fields: [code]Vector2i size_px[/code] - font size in pixels, [code]float viewport_oversampling[/code] - viewport oversampling factor, [code]int glyphs[/code] - number of rendered glyphs, [code]int textures[/code] - number of used textures, [code]int textures_size[/code] - size of texture data in bytes.
```

AThousandShips commented on doc/classes/TextServerExtension.xml:
```diff
@@ -1315,6 +1332,14 @@
+			<param index="0" name="oversampling" type="float" />
+			<description>
+				[b]Required.[/b]
+				Increases reference count of the specified oversampling level. This method is called by [Viewport], and should not be used directly.
```
```suggestion
				Increases the reference count of the specified oversampling level. This method is called by [Viewport], and should not be used directly.
```

AThousandShips commented on doc/classes/TextServerExtension.xml:
```diff
@@ -2109,5 +2136,13 @@
+			<param index="0" name="oversampling" type="float" />
+			<description>
+				[b]Required.[/b]
+				Decreases reference count of the specified oversampling level, and frees font cache for oversampling level when reference count reaches zero. This method is called by [Viewport], and should not be used directly.
```
```suggestion
				Decreases the reference count of the specified oversampling level, and frees the font cache for oversampling level when the reference count reaches zero. This method is called by [Viewport], and should not be used directly.
```

AThousandShips commented on scene/resources/text_line.compat.inc:
```diff
@@ -0,0 +1,46 @@
+#ifndef DISABLE_DEPRECATED
+
+void TextLine::_draw_bind_compat_104872(RID p_canvas, const Vector2 &p_pos, const Color &p_color) const {
+	draw(p_canvas, p_pos, p_color);
```
```suggestion
	draw(p_canvas, p_pos, p_color, 0.0);
```
Pass default argument explicitly

AThousandShips commented on scene/resources/text_line.compat.inc:
```diff
@@ -0,0 +1,46 @@
+}
+
+void TextLine::_draw_outline_bind_compat_104872(RID p_canvas, const Vector2 &p_pos, int p_outline_size, const Color &p_color) const {
+	draw_outline(p_canvas, p_pos, p_outline_size, p_color);
```
```suggestion
	draw_outline(p_canvas, p_pos, p_outline_size, p_color, 0.0);
```

AThousandShips commented on scene/resources/text_paragraph.compat.inc:
```diff
@@ -0,0 +1,68 @@
+#ifndef DISABLE_DEPRECATED
+
+void TextParagraph::_draw_bind_compat_104872(RID p_canvas, const Vector2 &p_pos, const Color &p_color, const Color &p_dc_color) const {
+	draw(p_canvas, p_pos, p_color, p_dc_color);
```
```suggestion
	draw(p_canvas, p_pos, p_color, p_dc_color, 0.0);
```

AThousandShips commented on scene/resources/text_paragraph.compat.inc:
```diff
@@ -0,0 +1,68 @@
+}
+
+void TextParagraph::_draw_outline_bind_compat_104872(RID p_canvas, const Vector2 &p_pos, int p_outline_size, const Color &p_color, const Color &p_dc_color) const {
+	draw_outline(p_canvas, p_pos, p_outline_size, p_color, p_dc_color);
```
```suggestion
	draw_outline(p_canvas, p_pos, p_outline_size, p_color, p_dc_color, 0.0);
```

AThousandShips commented on scene/resources/text_paragraph.compat.inc:
```diff
@@ -0,0 +1,68 @@
+}
+
+void TextParagraph::_draw_line_bind_compat_104872(RID p_canvas, const Vector2 &p_pos, int p_line, const Color &p_color) const {
+	draw_line(p_canvas, p_pos, p_line, p_color);
```
```suggestion
	draw_line(p_canvas, p_pos, p_line, p_color, 0.0);
```

Repiteo commented:
Thanks!

darksylinc commented:
This PR completely broke Godot on Xubuntu:

![image](https://github.com/user-attachments/assets/db459aee-2645-47df-86bd-e25e60962782)

That's all I get when opening the Editor.
